test/jdk/java/util/regex/RegExTest.java

```diff
@@ -4548,145 +3894,126 @@ private static void unicodeCharacterNameTest() throws Exception {
     }
 
-    private static void horizontalAndVerticalWSTest() throws Exception {
+    @Test
+    public static void horizontalAndVerticalWSTest() {
         String hws = new String (new char[] {
                                      0x09, 0x20, 0xa0, 0x1680, 0x180e,
                                      0x2000, 0x2001, 0x2002, 0x2003, 0x2004, 0x2005,
                                      0x2006, 0x2007, 0x2008, 0x2009, 0x200a,
                                      0x202f, 0x205f, 0x3000 });
         String vws = new String (new char[] {
                                      0x0a, 0x0b, 0x0c, 0x0d, 0x85, 0x2028, 0x2029 });
-        if (!Pattern.compile("\\h+").matcher(hws).matches() ||
-            !Pattern.compile("[\\h]+").matcher(hws).matches())
-            failCount++;
-        if (Pattern.compile("\\H").matcher(hws).find() ||
-            Pattern.compile("[\\H]").matcher(hws).find())
-            failCount++;
-        if (!Pattern.compile("\\v+").matcher(vws).matches() ||
-            !Pattern.compile("[\\v]+").matcher(vws).matches())
-            failCount++;
-        if (Pattern.compile("\\V").matcher(vws).find() ||
-            Pattern.compile("[\\V]").matcher(vws).find())
-            failCount++;
+        assertTrue(Pattern.compile("\\h+").matcher(hws).matches() &&
+                   Pattern.compile("[\\h]+").matcher(hws).matches());
+        assertTrue(!Pattern.compile("\\H").matcher(hws).find() &&
+                   !Pattern.compile("[\\H]").matcher(hws).find());
+        assertTrue(Pattern.compile("\\v+").matcher(vws).matches() &&
+                   Pattern.compile("[\\v]+").matcher(vws).matches());
+        assertTrue(!Pattern.compile("\\V").matcher(vws).find() &&
+                   !Pattern.compile("[\\V]").matcher(vws).find());
         String prefix = "abcd";
         String suffix = "efgh";
         String ng = "A";
         for (int i = 0; i < hws.length(); i++) {
             String c = String.valueOf(hws.charAt(i));
             Matcher m = Pattern.compile("\\h").matcher(prefix + c + suffix);
-            if (!m.find() || !c.equals(m.group()))
-                failCount++;
+            assertTrue(m.find() && c.equals(m.group()));
             m = Pattern.compile("[\\h]").matcher(prefix + c + suffix);
-            if (!m.find() || !c.equals(m.group()))
-                failCount++;
-
-            m = Pattern.compile("\\H").matcher(hws.substring(0, i) + ng + hws.substring(i));
-            if (!m.find() || !ng.equals(m.group()))
-                failCount++;
-            m = Pattern.compile("[\\H]").matcher(hws.substring(0, i) + ng + hws.substring(i));
-            if (!m.find() || !ng.equals(m.group()))
-                failCount++;
+            assertTrue(m.find() && c.equals(m.group()));
+
+            String matcherSubstring = hws.substring(0, i) + ng + hws.substring(i);
+
+            m = Pattern.compile("\\H").matcher(matcherSubstring);
+            assertTrue(m.find() || ng.equals(m.group()));
```

Comment:
Should this be:
`assertTrue(m.find() && ng.equals(m.group()));`

Reply:
Good catch. De Morgan's mistake